### DISCUSSION TITLE
Half-cell storage operators?

### DISCUSSION THREAD

**Discussion by SirWumpus** [ForthHub/discussion#144] [General]
With 8-bit CPUs with 16 bit words, a half-word is typically an octet accessible with `C@` and `C!` and the full word with `@` and `!`.

With larger CPU word sizes, ie. 64 bit (maybe one day 128 bit), is it worth considering standardising some basic support for half word / cell size storage operators (assuming cells match the machine CPU word size)?  In particular for mapping structures to regions of memory or handling protocol objects where some fields are smaller than a cell?  Something like `H@` and `H!`?

**Comment by bshepherdson** [comment]
I think many Forths support this, but there's little standardization of names (just as there is in assembly terminology, "words" vs. "longs" vs. "halfwords").

Gforth has `W@`/`W!` for words = 16-bit values and `L@/L!` for longs = 32-bit values. That uses Intel-ish naming conventions and is probably a good naming scheme to borrow, unless you're targeting a specific machine with different naming.

As you say, these are useful for reading and writing structures not defined in Forth. Of course any application can roll its own from `c@` and c!`, but if the system provides these words it should be able to use more efficient CPU instructions.

**Comment by ruv** [comment]
This problem was mentioned on the standard meeting in February: Memory operators 8, 16, 32, 64 bits - signed/unsigned.

It's not an easy problem.

1. Should it support reading of signed integers? Reading from a region that is less than a cell can be unsigned or signed. In the latter case it should propagate the sign — assuming two's complement format, it should copy the sign bit (MSB) from the region to the more significant bits of the cell.
2. Should it support different endianness (bytes order)?
3. Should it support different endianness in double-cell reading/writing?
4. Should it support reading from a region that is greater than a cell into a cell? Ditto for writing operations. What to do on overflow (it cannot be ignored)?
5. If we dedicate one operation to one word, then combination of all possible variants leads to a very big bunch of words. Can we specify one operation by several words, so the number of different words is not so big?  Is it worth using a recognizable format for that?

**Reply by SirWumpus** [reply]
> Should it support reading of signed integers?

I was thinking more like how the M68000 read byte, word, or long values.  When reading there was no consideration of sign as I recall, treat it as unsigned / as-is.  I suspect the issue of sign extension would then be left to the caller to handle.  

> Should it support different endianness (bytes order)?

Endian when its handled by machine instructions would not be an issue.  Software emulation would have to deal with such and I'd expect behaviour as though it was read by using hardware instructions (if they had any for such).  

> Should it support different endianness in double-cell reading/writing?

I was thinking memory operations smaller than a cell.  Double-cell means two cells, so you treat them as two cells, not combined to be a super-cell (hurricanes not with standing).

> Should it support reading from a region that is greater than a cell into a cell?

I'd think that was a "no".  The question focuses on reading regions bigger than an octet and smaller than a cell.   Not sure how reading reading a supercell into a cell would ever work.  The operators should just read subsize cell memory without interpretatio1n.  

> If we dedicate one operation to one word, then combination of all possible variants leads to a very big bunch of words. ...

It can get messy quick.  If one opts to individual words, then a format is probably wise `M8@` `M8!` ... `M16@` ... `M64@`.  Or given how some words already prefix `@/!` with a value, eg `2@`,  maybe flip it `@8 .. @16 .. @64` etc.  Or to reduce the number of words a more generic option like:  `foo 8  @@`  `bar  32  @@`  or `$1234 xyzzy 16 !!` ,   `@@ ( addr b  -- ub )` and `!! ( ub addr b -- )`.

**Reply by ruv** [reply]
Concerning endianness — I mean that sometimes you have to read or write a number in specific endianness, regardless what endianness is used in the Forth system.

I think, this set of operation should include both the octet-size operations and cell-size operations. I.e., a 32-bit Forth system should include `L@`, and `L!` (probably as synonyms for `@` and `!`). Otherwise a program is forced to test its environment and use `@` on a 32-bit Forth system, and `L@` on a 64-bit Forth system.

Concerning octets — `c@` does not read an octet, but an address unit now, which can be 8, 16, or even 32 bits.
And interesting question is how to address octets if the address unit is more than octet. If in a Forth system the address unit is 32 bit, the system cannot provide the words `B@` and `W@`, can it?

**Reply by ruv** [reply]
> how to address octets if the address unit is more than octet

JavaScript employs the "DataView" interface, which allows to read/write data at any byte-offset, and without alignment constraints.

As a simple solution, there can be a single data view, you set it's starting address, and the words accept offset from this starting address in bytes, regardless the address unit of the Forth system and its alignment constraints.

The same approach can be used to access individual bits too.

**Comment by MitchBradley** [comment]
This wheel was invented and perfected a long time ago.  Some historical / common practice context:

I was among the first people to use Forth on 32-bit machines.  In 1982, I modified Mike Perry's 16-bit 68000 Forth to use 32-bit stacks, cells and addresses, then used it extensively for diagnostics on Sun workstations.  In those days, I regularly attended nearly every Forth conference, including FIG meetings, FORML, and Rochester.  Nobody else was doing 32-bit Forth except Bill Sebok who was using Forth on a VAX.  In fact most people were completely uninterested in 32-bit systems. Bill and I compared notes and agreed on a common set of names for various width-dependent operations, including:

  16-bit: W!, W@ (unsigned), <W@  (sign extend to a cell)
  32-bit: L!, L@ (unsigned), <L@ (sign extend to a cell, anticipating future larger cell sizes)

We also had words like /W (number of address units of a 16-bit word, i.e. 2 for byte-addressed machines) and /L (..4), plus convenience operators like WA1+ (typically same as 2+), WA+ ( adr #words -- adr' ), and similar for longs. /W* was like 2* etc.

N instead of W or L meant "normal".  The "N" notation was eventually deprecated many years later when the ANS committee adopted the "CELL" nomenclature.

My early systems led to a number of my public domain/shareware/open source Forth systems including ForthMacs, ForthMon, CForth, and eventually Open Firmware. I have used those W and L words in all of those systems, including Open Firmware which was deployed on millions of commercial machines in its heyday.

In addition, Open Firmware had explicit-endian versions:  LE-W@, LE-W!, BE-W@, BE-W!, LE-L@, LE-L!, BE-L@, BE-L! .  They get a lot of use for accessing items in externally-defined data structures like network packets and descriptors that are passed to and from "smart" devices.

When 64-bit systems appeared, the W and L prefixes were extended to include X, as in X@ .

It can get even more complicated when you consider alignment and granularity.  In some cases, you absolutely must read or write the data in a single machine operation, as when accessing device registers.  For that, Open Firmware has e.g. "RL!", R meaning "register", guaranteeing that that operation will not be split into pieces.  That one is tricky to implement since the hardware operation can be bus-dependent, both for alignment and endianness; Open Firmware handles that by making the "R" works operate in the context of the parent bus.

In other cases, you need to be able to access a memory data structure that might not respect the same alignment restrictions as the CPU you are running on.  For example you might have to decode a network packet with a 32-bit field that is not on a 32-bit boundary, while running on a CPU whose native 32-bit fetch operation will fault if the address is not a multiple of 4.  The Open Firmware BE- and LE- data access words make appropriate guarantees so you can access such data without faulting.

It seems like a lot of words, but every one of them serves an important purpose if you are trying to solve real-world problems without pretending that the world is simpler than it really is.

**Reply by ruv** [reply]
> 16-bit: `W!`, `W@` (unsigned), `<W@` (sign extend to a cell)
> 32-bit: `L!`, `L@` (unsigned), `<L@` (sign extend to a cell, anticipating future larger cell sizes)

Now I would prefer the names `SW@`, `SL@` for the words that read signed numbers.
Also, we probably need words to access octets (bytes), since a char can have another size (?).

> Open Firmware had explicit-endian versions: `LE-W@`, `LE-W!`, `BE-W@`, `BE-W!`, `LE-L@`, `LE-L!`, `BE-L@`, `BE-L!` . They get a lot of use for accessing items in externally-defined data structures like network packets and descriptors that are passed to and from "smart" devices.

What formats were supported in your systems for signed number representation?
Is it enough to support only the "two's complement" format to access externally-defined data nowadays?

What do you think, is it better to have separate words to propagate the sign, something like `W>S ( +n|u -- n )`  and `L>S ( +n|u -- n )`, instead of the words to read signed numbers (as `SW@`, `LE-SW@`, `BE-SW@`, etc)?

**Comment by jwoehr** [comment]
On Tue, Mar 7, 2023 at 5:19 PM Mitch Bradley ***@***.***>
wrote:

>  Nobody else was doing 32-bit Forth except Bill Sebok who was using Forth
> on a VAX. In fact most people were completely uninterested in 32-bit
> systems.


They weren't precisely "uninterested", Mitch ... Not many of us had access
to the upscale toys you had access to! Those were the days, right!?

"Have you noticed that each generation of Sun workstation is increasingly
optimized to run the CAD software necessary to design a Sun workstation?"
- Mitch Bradley, *circa* 1990.

😁😁😁

-- 

*JACK WOEHR, IBM Champion 2021-2023
<https://www.credly.com/users/jack-woehr/badges>*

*SENIOR IBM i ENGINEER*

  303.847.8442
  ***@***.***
  stay connected
<https://www.linkedin.com/company/absolute-performance-inc./>

<https://www.absolute-performance.com/>
 www.absolute-performance.com

NON-DISCLOSURE NOTICE:  This communication including any and all
attachments is for the intended recipient(s) only and may contain

confidential and privileged information.  If you are not the intended
recipient of this communication, any disclosure, copying further

distribution or use of this communication is prohibited.  If you received
this communication in error, please contact the sender and

delete/destroy all copies of this communication immediately.

**Comment by MitchBradley** [comment]
Re W>S and L>S:  If you admit the N prefix for "normal" i.e. cells, you can have W>N and L>N.  Similarly you can have BE-N@ etc - or you could just say LE@ etc.  It all becomes regular.  That is what Open Firmware does.

From the 32-bit version of the ARM kernel:
```
code s->l  ( n -- l )  c;
code l->n  ( l -- n )  c;
code n->a  ( n -- a )  c;
code l->w  ( l -- w )  mov tos,tos,lsl #16  mov tos,tos,lsr #16  c;
code n->w  ( n -- w )  mov tos,tos,lsl #16  mov tos,tos,lsr #16  c;
code w->n  ( n -- w )  mov tos,tos,lsl #16  mov tos,tos,asr #16  c;  \ Sign extend
```

As C@ and C! became divorced from octets, I started using B for byte as a prefix.  Thus Open Firmware's "register access words" for bytes are RB@ and RB!. 

Another thing that Open Firmware has is splits, joins, and flips:

wbsplit ( w -- b.low b.high)
bwjoin ( b.low b.high -- w )
lbsplit ( l -- b.low b.lowmid b.highmid b.high )
bljoin ( b.low b.lowmid b.highmid b.high -- l )
wflip ( w1 -- w2 )
lflip (l1 -- l2 )

They are surprisingly useful in a number of cases, especially for device drivers.

As a large system, Open Firmware included a lot of words that might be "too much" for a minimalist, but in practice they were used quite a lot, so they paid for themselves.

My argument to minimalists is this: Even if you don't end up needing or implementing some words in a given system, there is still great value to specifying names and behaviors for when you or someone else does need them.

**Reply by ruv** [reply]
> Re `W>S` and `L>S`: If you admit the N prefix for "normal" i.e. cells, you can have `W>N` and `L>N`.

My idea is that these words follow the standard [word](https://forth-standard.org/standard/double/DtoS) `D>S ( d -- n )` naming, where *S* means _signed_ number (single-cell).

> Similarly you can have `BE-N@` etc - or you could just say `LE@` etc. It all becomes regular.

Are the words like `LE@`, `BE@` really useful/demanded?  They assume that endianness is specified, but the region width is not specified (which is the cell size, which is system-defined).

**Comment by MitchBradley** [comment]
> Are the words like LE@, BE@ really useful/demanded? They assume that endianness is specified, but the region width is not specified (which is the cell size, which is system-defined).

That is probably correct.

**Comment by spelc** [comment]
This topic has been worked on many times in standards meetings from ANS onwards. There has never been a proposal that has achieved enough support. There are essentially two approaches; 1) build a tool set for constructing these words, or 2) provide a (perhaps extensible) set of words. 

The first solution doesn't solve the problem of how to name things, e.g. @LE, @BE and @N for little-endian, big-endian and network order. For the cross-compiler author it's even worse as you need @C(T) @D(T) and @J(T) for target code space, target data space, and real target access by JTAG. In the second solution you end up proposing an extensible solution or a huge number of words. 

Despite persistent explanation that you do not need to implement everything to be standards compliant, the message has never reached popular support.

Those of us who write or maintain cross compilers regularly simply talk to each other!

**Reply by spelc** [reply]
There are two proposals on this topic at [forth-standard.org](https://forth-standard.org/) by myself and Leon Wagner ([1](https://forth-standard.org/proposals/16-bit-memory-access), [2](https://forth-standard.org/proposals/32-bit-memory-operators)).
